modules/rtm/R/bayestools.R

```diff
@@ -0,0 +1,168 @@
+    prior_df <- prior_df_all[prior_df_all[['param_name']] %in% use_names,]
+    prior <- PEcAn.assim.batch::pda.create.btprior(prior_df)
+    return(prior)
+}
+
+#' Perform Bayesian inversion using BayesianTools package
+#' 
+#' Use samplers from the BayesianTools package to fit models to data. Like 
+#' `invert.auto`, this will continue to run until convergence is achieved 
+#' (based on Gelman diagnostic) _and_ the result has enough samples (as 
+#' specified by the user; see Details).
+#' 
+#' @details `custom_settings` is a list of lists, containing the following:
+#'  * `common` -- BayesianTools settings common to both the initial and subsequent samples.
+#'  * `init` -- BayesianTools settings for just the first round of sampling. 
+#'  This is most common for the initial number of iterations, which is the 
+#'  minimum expected for convergence.
+#'  * `loop` -- BayesianTools settings for iterations inside the convergence 
+#'  checking `while` loop. This is most commonly for setting a smaller 
+#'  iteration count than in `init`.
+#'  * `other` -- Miscellaneous (non-BayesianTools) settings, including:
+#'      - `sampler` -- String describing which sampler to use. Default is `DEzs`
+#'      - `use_mpsrf` -- Use the multivariate PSRF to check convergence. 
+#'      Default is `FALSE` because it may be an excessively conservative 
+#'      diagnostic.
+#'      - `min_samp` -- Minimum number of samples after burnin before stopping.
+#' 
+#' See the BayesianTools sampler documentation for what can go in the `BayesianTools` settings lists.
+#' @param observed Vector of observations
+#' @param model Function called by log-likelihood. Must be `function(params)` 
+#' and return a vector equal to `length(observed)` or `nrow(observed)`.
+#' @param prior BayesianTools prior object.
+#' @param custom_settings Nested settings list. See Details.
+#' @export
+invert_bt <- function(observed, model, prior, custom_settings = list()) {
+
+    default_settings <- list(common = list(), 
+                             init = list(iterations = 10000),
+                             loop = list(iterations = 2000),
+                             other = list(sampler = 'DEzs', 
+                                          use_mpsrf = FALSE,
+                                          min_samp = 1000))
+
+    if (length(custom_settings) > 0) {
+        for (s in seq_along(default_settings)) {
+            s_name <- names(default_settings)[s]
+            if (s_name %in% names(custom_settings)) {
+                settings[[s_name]] <- modifyList(default_settings[[s_name]], 
+                                                 custom_settings[[s_name]])
+            } else {
+                settings[[s_name]] <- default_settings[[s_name]]
+            }
+        } 
+    } else {
+        settings <- default_settings
+    }
+
+    use_mpsrf <- settings[['other']][['use_mpsrf']]
+    min_samp <- settings[['other']][['min_samp']]
+    lag.max <- settings[['other']][['lag.max']]
```

Comment:
Default is calculated the same way as in the `acf` function -- something like 10 x log(n). Perhaps `settings` is confusing -- that's really coming from the `custom_settings`. Though because it's a list of lists, there's no reason you couldn't store settings in XML, YAML, etc. and then read them in one line and pass them here.